mono44: Init at 4.4.1.0

vrthra opened the pull request:
###### Motivation for this change
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bjornfor commented:
I guess I'm more nitpick'y about these things than many others here, but I think you should either split the commit it two, to separate the version bump from the other clean-ups, or (at least) make a comment about the other non-functional changes that are included. That makes it easier for reviews.

vrthra commented:
@bjornfor Thanks, I appreciate any feedback on my contributions. I am still learning the ropes in this community.

bjornfor commented:
@vrthra: Thanks! (Building locally...)

bjornfor commented:
I found that including this PR in its current state would break the build for the following packages:

```
pdfmod-0.9.1
Newtonsoft.Json-6.0.8
EventStore-3.5.0
Paket-1.18.2
FSharp.AutoComplete-0.18.2
hyena-0.5
mono-zeroconf-0.9.0
nant-2015-11-15
gdata-sharp-2.2.0.0
ExtCore-0.8.46
FSharp.Compiler.Service-0.0.90
openra-20160508
fstar-0.9.2.0
Deedle-1.2.0
monodevelop-5.9.4.5
```

(List obtained with ```nox-review wip ...``` and ```nix-build --keep-going ...``` with and without this PR.)

vrthra commented:
@bjornfor  Thank you, I will try and fix these and get back.

bjornfor commented:
Worst case, mono is so unstable that we have to have mono-4.0, mono-4.4...

vrthra commented:
There seems to be multiple issues, with a renaming of API libraries from `<number>` to `<number>-api` and also, warnings changed to default to errors. Would it be OK to leave `mono-4.0` as default, and add `mono-4.4` as a new entry?

bjornfor commented:
Yes, that seems best (IMHO).

vrthra commented:
@bjornfor Updated with just the `mono44` commit as a new entry.

bjornfor commented:
@vrthra: Hm, I don't like the duplicated code. (There were no functional changes needed for 4.4, right?) Can you create a generic.nix, 4.0.nix and 4.4.nix? You can see nixops is using this method (except the versions are called "default" and "unstable", but that doesn't apply here): https://github.com/NixOS/nixpkgs/tree/master/pkgs/tools/package-management/nixops

vrthra commented:
@bjornfor I have updated the build configurations after refactoring. On `nix-shell -p nox --run "nox-review wip"`, `fstar, boo, and banshee` fails. However, they seem to fail on building in the current master too.

vrthra committed:
boo: mark as broken

With mono-4.0.4.1, the build fails with 1 non-fatal errors.

vrthra committed:
ue4: mark as broken

The source code is no longer available as the repository
https://github.com/EpicGames/UnrealEngine has been deleted.

bjornfor commented on pkgs/development/compilers/boo/default.nix:
```diff
@@ -41,5 +41,7 @@ in stdenv.mkDerivation rec {
   meta = with stdenv.lib; {
     description = "The Boo Programming Language";
     platforms = platforms.linux;
+    hydraPlatforms = stdenv.lib.platforms.none;
```
I don't think setting hydraPlatforms does anything useful here. (Your "meta = broken" is all that's needed.)

bjornfor commented on pkgs/games/ue4/default.nix:
```diff
@@ -78,5 +78,7 @@ stdenv.mkDerivation rec {
     license = stdenv.lib.licenses.unfree;
     platforms = stdenv.lib.platforms.linux;
     maintainers = [ stdenv.lib.maintainers.puffnfresh ];
+    hydraPlatforms = stdenv.lib.platforms.none;
```
hydraPlatforms not needed.

bjornfor commented:
More nitpicks: It'd be a smaller diff, and easier review, if you make the generic.nix file first, instead of introducing a full copy of the expression for 4.4 and then _later_ merge the two files.

vrthra committed:
mono: refactor and cleanup

Refactor and cleanup in preparation for mono44

vrthra committed:
mono40: refactor to generic 40

vrthra committed:
mono44: init at 4.4.1.0

vrthra commented:
@bjornfor I have updated the build as you suggest, with generic first, and then 4.4

bjornfor commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4805,6 +4805,11 @@ in
     inherit (darwin.apple_sdk.frameworks) Foundation;
   };
 
+  mono44 = callPackage ../development/compilers/mono/4.4.nix {
```
I think we should make mono44 "lowPrio" for now, so that ```nix-env -i mono``` installs the same version as ```nix-env -iA mono```. (Without lowPrio, ```nix-env -i mono``` installs mono-4.4, which may seem odd when ```nix-env -iA mono``` installs mono-4.0.)

vrthra replied:
@bjornfor updated to lowprio

bjornfor commented:
@vrthra: Excellent! Testing locally, will push when done testing.

bjornfor commented:
Pushed to master (a783a5704f8ec09041f24bb356812e1c7494c0d5 and parents). Thanks!
